python/benchmark/benchmark/bench_dbscan.py

```diff
@@ -0,0 +1,283 @@
+                    array(*input_cols).alias("features_array"), output_col
+                )
+            elif is_vector_col:
+                df_for_scoring = transformed_df.select(
+                    vector_to_array(col(feature_col)), output_col
+                )
+
+        if num_cpus > 0:
+            from pyspark.ml.clustering import KMeans as SparkKMeans
+
+            assert num_gpus <= 0
+            if is_array_col:
+                vector_df = train_df.select(
+                    array_to_vector(train_df[first_col]).alias(first_col)
+                )
+            elif not is_vector_col:
+                vector_assembler = VectorAssembler(outputCol="features").setInputCols(
+                    input_cols
+                )
+                vector_df = vector_assembler.transform(train_df).drop(*input_cols)
+                first_col = "features"
+            else:
+                vector_df = train_df
+
+            if not no_cache:
+
+                def cpu_cache_df(df: DataFrame) -> DataFrame:
+                    df = df.cache()
+                    df.count()
+                    return df
+
+                vector_df, prepare_time = with_benchmark(
+                    "prepare dataset", lambda: cpu_cache_df(vector_df)
+                )
+
+            params = self.class_params
+            print(f"Passing {params} to KMeans")
+
+            cpu_estimator = (
+                SparkKMeans(**params)
+                .setFeaturesCol(first_col)
+                .setPredictionCol(output_col)
+            )
+
+            cpu_model, fit_time = with_benchmark(
+                "cpu fit", lambda: cpu_estimator.fit(vector_df)
+            )
+
+            print(
+                f"spark ML: iterations: {cpu_model.summary.numIter}, inertia: {cpu_model.summary.trainingCost}"
+            )
+
+            def cpu_transform(df: DataFrame) -> None:
+                transformed_df = cpu_model.transform(df)
+                transformed_df.agg(sum(output_col)).collect()
+                return transformed_df
+
+            transformed_df, transform_time = with_benchmark(
+                "cpu transform", lambda: cpu_transform(vector_df)
+            )
+
+            total_time = time.time() - func_start_time
+            print(f"cpu total took: {total_time} sec")
+
+            feature_col = first_col
+            df_for_scoring = transformed_df.select(
+                vector_to_array(col(feature_col)).alias(feature_col), output_col
+            )
+            cluster_centers = cpu_model.clusterCenters()
+
+        # either cpu or gpu mode is run, not both in same run
+        score = self.score(df_for_scoring, feature_col, output_col)
+        # note: seems that inertia matches score at iterations-1
+        print(f"score: {score}")
+
+        result = {
+            "fit_time": fit_time,
+            "transform_time": transform_time,
+            "total_time": total_time,
+            "score": score,
```

Comment:
use "silhoutte score" to be more specific in the report.csv?

Reply:
This seems like the common output of a "score" parameter where each algo benchmark has different interpretation.